src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs

```diff
@@ -209,14 +222,74 @@ private void onBlockConnected(ChainedHeaderBlock chBlock)
         }
 
         // TODO subscribe to signals
-        private void onBlockDisconnected(ChainedHeaderBlock chBlock)
+        // TODO make private when signaler PR is merged
+        public void onBlockDisconnected(ChainedHeaderBlock chBlock)
         {
-            // TODO it should handle reorgs and revert votes that were applied.
+            byte[] rawVotingData = this.votingDataEncoder.ExtractRawVotingData(chBlock.Block.Transactions[0]);
+
+            List<VotingData> votingDataList = this.votingDataEncoder.Decode(rawVotingData);
+            votingDataList.Reverse();
+
+            if (rawVotingData == null)
+            {
+                this.logger.LogTrace("(-)[NO_VOTING_DATA]");
+                return;
+            }
+
+            lock (this.locker)
+            {
+                foreach (VotingData votingData in votingDataList)
+                {
+                    // Poll that was finished in the block being disconnected.
+                    Poll pollToRevert = this.finishedPolls.SingleOrDefault(x => x.PollAppliedBlockHash == chBlock.Block.GetHash() && x.VotingData == votingData);
```

Comment:
`chBlock.Block.GetHash() ` is the `ChainedHeader.HashBlock`

Reply:
yeah
and both of them will returned cached value